ext/xml/xml.c

```diff
@@ -1055,168 +1069,194 @@ PHP_FUNCTION(xml_set_object)
+	zend_fcall_info_cache end_fcc = {0};
+	zend_string *method_name = NULL;
+	zend_string *method_name2 = NULL;
+
+	// TODO: cover trampolines with tests, as the !ZEND_FCC_INITIALIZED branches are never executed right now
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "Of!f!", &pind, xml_parser_ce, &start_fci, &start_fcc, &end_fci, &end_fcc) == SUCCESS) {
+		parser = Z_XMLPARSER_P(pind);
+		if (ZEND_FCI_INITIALIZED(start_fci) && !ZEND_FCC_INITIALIZED(start_fcc)) {
+			zend_is_callable_ex(&start_fci.function_name, NULL, IS_CALLABLE_SUPPRESS_DEPRECATIONS, NULL, &start_fcc, NULL);
+			/* Call trampoline has been cleared by zpp. Refetch it, because we want to deal
+			 * with it ourselves. It is important that it is not refetched on every call,
+			 * because calls may occur from different scopes. */
+		}
+		if (ZEND_FCI_INITIALIZED(end_fci) && !ZEND_FCC_INITIALIZED(end_fcc)) {
+			zend_is_callable_ex(&end_fci.function_name, NULL, IS_CALLABLE_SUPPRESS_DEPRECATIONS, NULL, &end_fcc, NULL);
+			/* Call trampoline has been cleared by zpp. Refetch it, because we want to deal
+			 * with it ourselves. It is important that it is not refetched on every call,
+			 * because calls may occur from different scopes. */
+		}
+	} else if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "Of!S", &pind, xml_parser_ce, &start_fci, &start_fcc, &method_name) == SUCCESS) {
+		parser = Z_XMLPARSER_P(pind);
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
-		RETURN_THROWS();
-	}
+		bool status = php_xml_check_string_method_arg(3, parser, method_name, &start_fcc);
+		if (status == false) {
+			RETURN_FALSE;
+		}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->notationDeclHandler, hdl);
-	XML_SetNotationDeclHandler(parser->parser, _xml_notationDeclHandler);
+		if (ZEND_FCI_INITIALIZED(start_fci) && !ZEND_FCC_INITIALIZED(start_fcc)) {
+			zend_is_callable_ex(&start_fci.function_name, NULL, IS_CALLABLE_SUPPRESS_DEPRECATIONS, NULL, &start_fcc, NULL);
+			/* Call trampoline has been cleared by zpp. Refetch it, because we want to deal
+			 * with it ourselves. It is important that it is not refetched on every call,
+			 * because calls may occur from different scopes. */
+		}
+	} else if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OSf!", &pind, xml_parser_ce, &method_name, &end_fci, &end_fcc) == SUCCESS) {
+		parser = Z_XMLPARSER_P(pind);
 
-	RETURN_TRUE;
-}
-/* }}} */
+		bool status = php_xml_check_string_method_arg(2, parser, method_name, &start_fcc);
+		if (status == false) {
+			RETURN_FALSE;
+		}
 
-/* {{{ Set up external entity reference handler */
-PHP_FUNCTION(xml_set_external_entity_ref_handler)
-{
-	xml_parser *parser;
-	zval *pind, *hdl;
+		if (ZEND_FCI_INITIALIZED(end_fci) && !ZEND_FCC_INITIALIZED(end_fcc)) {
+			zend_is_callable_ex(&end_fci.function_name, NULL, IS_CALLABLE_SUPPRESS_DEPRECATIONS, NULL, &end_fcc, NULL);
```

Comment:
Silly nit: I wonder if we can somehow extract this common trampoline handling code, I'm having many déjà vu's :p

Reply:
Yeah, this piece of code would definitely be nice to be extracted

Reply:
> Yeah, this piece of code would definitely be nice to be extracted

See https://github.com/php/php-src/pull/12375, as the issue is the trampoline being freed by ZPP.